src/cast_core/src/CollectionScanner.cpp

```diff
@@ -227,9 +241,117 @@ void countScan(CollectionScanner::PhaseConfig* config,
     statTracker.success();
 }
 
+void pointInTimeScan(mongocxx::pool::entry& client,
+                     CollectionScanner::PhaseConfig* config,
+                     std::optional<bsoncxx::types::b_timestamp> readClusterTime) {
+    int64_t docCount = 0;
+    int64_t scanSize = 0;
+    bool scanFinished = false;
+    mongocxx::options::client_session sessionOption;
+    sessionOption.causal_consistency(false);
+    mongocxx::client_session session = client->start_session(sessionOption);
+    if (!readClusterTime) {
+        client->database("admin").run_command(session, make_document(kvp("ping", 1)));
+        readClusterTime = session.operation_time();
+    }
+    bsoncxx::document::value readConcern = bson_stream::document{}
+        << "level"
+        << "snapshot"
+        << "atClusterTime" << readClusterTime.value() << bson_stream::finalize;
+    auto filter =
+        config->filterExpr ? config->filterExpr->evaluate() : bsoncxx::document::view_or_value{};
+    for (auto database : config->databases) {
+        auto collectionNames = [&] {
+            if (config->queryCollectionList) {
+                return database.list_collection_names({});
+            } else {
+                return config->collectionNames;
+            }
+        }();
+        for (auto& collection : collectionNames) {
+            /*
+             * Try-catch this as the collection may have been deleted.
+             * You can still do a find but it'll throw an exception when we iterate.
+             */
+            try {
+                bsoncxx::document::value findCmd = bson_stream::document{}
+                    << "find" << collection << "filter" << filter << "readConcern"
+                    << readConcern.view() << bson_stream::finalize;
+
+                auto findTracker = config->scanOperation.start();
+                auto res = database.run_command(session, findCmd.view());
```

Comment:
if we barf here (like it seems like we can based on your comment above) the tracker will be held open and we'll fail to record the data-point. I think just move the `try` opening statement to below the `.start` and in your catch block call .failure.

Reply:
Good catch! Done.